midend/flattenInterfaceStructs.h

```diff
@@ -0,0 +1,189 @@
+        const IR::Expression* root, cstring prefix);
+};
+
+/**
+ * Maintains a map between an original struct type and its
+ * replacement.
+ */
+struct NestedStructMap {
+    P4::ReferenceMap* refMap;
+    P4::TypeMap* typeMap;
+
+    ordered_map<const IR::Type*, StructTypeReplacement*> replacement;
+
+    NestedStructMap(P4::ReferenceMap* refMap, P4::TypeMap* typeMap):
+            refMap(refMap), typeMap(typeMap)
+    { CHECK_NULL(refMap); CHECK_NULL(typeMap); }
+    void createReplacement(const IR::Type_Struct* type);
+    StructTypeReplacement* getReplacement(const IR::Type* type) const
+    { return ::get(replacement, type); }
+    bool empty() const { return replacement.empty(); }
+};
+
+/**
+Find the types to replace and insert them in the nested struct map.
+This pass only looks at type arguments used in package instantiations.
+ */
+class FindTypesToReplace : public Inspector {
+    NestedStructMap* map;
+ public:
+    explicit FindTypesToReplace(NestedStructMap* map): map(map) {
+        setName("FindTypesToReplace");
+        CHECK_NULL(map);
+    }
+    bool preorder(const IR::Declaration_Instance* inst) override;
+};
+
+/**
+This pass transforms the type signatures of instantiated controls,
+parsers, and packages.  It does not transform methods, functions or
+actions.  It starts from package instantiations: every type argument
+that is a nested structure is replaced with "simpler" flat type.
+
+Should be run after the NestedStructs pass.
+
+struct S { bit b; }
+struct T { S s; }
+
+control c(inout T arg) {
+    apply {
+        ... arg.s.b ...
+    }
+}
+
+control proto<V>(inout V arg);
+package top<V>(proto<V> ctrl);
+
+top<T>(c()) main;
+
+This is transformed into:
+
+struct S { bit b; }
+struct T {
+   bit _s_b0;
+}
+
+control c(inout T arg) {
+    apply {
+        ... arg._s_b0; ...
+    }
+}
+
+top<TFlat>(c()) main;
+
+ */
+class ReplaceStructs : public Transform {
+    NestedStructMap* replacementMap;
+    std::map<const IR::Parameter*, StructTypeReplacement*> toReplace;
+
+ public:
+    explicit ReplaceStructs(NestedStructMap* sm): replacementMap(sm) {
+        CHECK_NULL(sm);
+        setName("ReplaceStructs");
+    }
+
+    const IR::Node* preorder(IR::P4Program* program) override;
+    const IR::Node* postorder(IR::Member* expression) override;
+    const IR::Node* preorder(IR::P4Parser* parser) override;
+    const IR::Node* preorder(IR::P4Control* control) override;
+    const IR::Node* postorder(IR::Type_Name* type) override;
+    const IR::Node* postorder(IR::Type_Struct* type) override;
+};
+
+class FlattenInterfaceStructs final : public PassManager {
+ public:
+    FlattenInterfaceStructs(ReferenceMap* refMap, TypeMap* typeMap) {
+        auto sm = new NestedStructMap(refMap, typeMap);
+        passes.push_back(new TypeChecking(refMap, typeMap));
+        passes.push_back(new FindTypesToReplace(sm));
+        passes.push_back(new ReplaceStructs(sm));
+        passes.push_back(new ClearTypeMap(typeMap));
```

Comment:
I assume you're clearing the typeMap because you've replaced the types. Which is the next pass that rebuilds it? I guess a different way of asking the question, aren't any invariants that are broken by where is `FlattenInterfaceStructs` inserted in the pass order?

Reply:
The contract enforced by the typeMap is that if an IR node is in a key in typeMap then the associated type is correct. This pass changes the types of some nodes, so it is easiest to clear the whole typeMap. In general, a pass runs type-checking if it needs to be sure that all nodes have types in the typeMap, but the type-checker skips over IR subtrees that are already in the typeMap. 

Due to the structure of our visitors, new IR nodes are created all the time, and it is quite frequent for nodes not to be in the typeMap, so typeChecking is run many times. It works as an integrity checker of the compiler. The alternative would be for each pass to carefully insert types for all new nodes created, but that is much more complicated than inserting nodes and leaving the type-checker compute their types.